File: packages/se-server/src/seKey.ts
import crypto from 'crypto';

/**
 * Stateless helpers for the SE.
 *
 * The SE's identity and one-time WOTS signing live in `./seIdentity` (RFC-008):
 * a root-identity anchor with leased one-time leaves. The legacy
 * `getPublicKeyHex` / `seSign` (a fixed index-0 key that did not match the
 * advertised digest) have been removed (AUD-003/AUD-025), as has the generic
 * flat-WOTS `wotsVerifyDigestAsync` — owner authentication is now root-bound
 * TreeKey verification (RFC-009, see `./ownerAuth`). What remains here are the
 * reclaim-tx encryption helpers.
 */

const RECLAIM_ENC_KEY_INFO = 'statechain-reclaim-tx-v1';

function getReclaimEncKey(seed: Uint8Array): Buffer {
  return crypto.createHmac('sha256', seed).update(RECLAIM_ENC_KEY_INFO).digest();
}

export function encryptReclaimTx(seed: Uint8Array, reclaimTxHex: string): string {
  const key = getReclaimEncKey(seed);
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const ct = Buffer.concat([cipher.update(reclaimTxHex, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `enc:${iv.toString('hex')}.${ct.toString('hex')}.${tag.toString('hex')}`;
}

export function decryptReclaimTx(seed: Uint8Array, enc: string): string {
  if (!enc.startsWith('enc:')) return enc;
  const parts = enc.slice(4).split('.');
  if (parts.length !== 3) throw new Error('Invalid encrypted reclaim tx format');
  const [ivHex, ctHex, tagHex] = parts;
  const key = getReclaimEncKey(seed);
  const iv = Buffer.from(ivHex, 'hex');
  const ct = Buffer.from(ctHex, 'hex');
  const tag = Buffer.from(tagHex, 'hex');
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(tag);
  return decipher.update(ct).toString('utf8') + decipher.final('utf8');
}

File: packages/se-server/go/sekey.go
package seserver

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/crypto/sha3"
)

const reclaimEncKeyInfo = "statechain-reclaim-tx-v1"

// ErrNotInteroperable marks the Go SE signing/identity surface as disabled
// (AUD-045). This package is NOT a WOTS-compatible SE: the TypeScript
// `@totemsdk/se-server` is the supported implementation. Until real
// interoperable WOTS (with cross-language known-answer tests) lands here, the
// signer/verifier fail closed rather than emitting placeholder cryptography
// that a peer could mistake for a valid SE signature.
var ErrNotInteroperable = errors.New("se-server(go): not an interoperable WOTS SE (AUD-045); use the TypeScript SE")

// getPublicKeyHex returns the (non-interoperable) legacy hash. It is retained
// only so the reference server still compiles; it is NOT the SE signing key and
// MUST NOT be treated as an SE identity. See ErrNotInteroperable.
func getPublicKeyHex(seed []byte) string {
	h := sha3.New256()
	h.Write(append(seed, 0, 0, 0, 0))
	return hex.EncodeToString(h.Sum(nil))
}

// seSign is disabled (AUD-045). The previous implementation returned an
// HMAC-SHA256 value, which is not a WOTS signature and must never be presented
// as one.
func seSign(seed, commitmentBytes []byte) ([]byte, error) {
	return nil, ErrNotInteroperable
}

// wotsVerifyDigest is disabled (AUD-045): never accept a non-WOTS signature.
func wotsVerifyDigest(sig, message, pkDigest []byte) bool {
	return false
}

func getReclaimEncKey(seed []byte) []byte {
	mac := hmac.New(sha256.New, seed)
	mac.Write([]byte(reclaimEncKeyInfo))
	return mac.Sum(nil)
}

func encryptReclaimTx(seed []byte, reclaimTxHex string) (string, error) {
	key := getReclaimEncKey(seed)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	iv := make([]byte, 12)
	if _, err := rand.Read(iv); err != nil {
		return "", err
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	ciphertext := aesgcm.Seal(nil, iv, []byte(reclaimTxHex), nil)
	return fmt.Sprintf("enc:%s.%s", hex.EncodeToString(iv), hex.EncodeToString(ciphertext)), nil
}

func decryptReclaimTx(seed []byte, enc string) (string, error) {
	if !strings.HasPrefix(enc, "enc:") {
		return enc, nil
	}

	parts := strings.SplitN(enc[4:], ".", 2)
	if len(parts) != 2 {
		return "", fmt.Errorf("invalid encrypted reclaim tx format")
	}

	iv, err := hex.DecodeString(parts[0])
	if err != nil {
		return "", err
	}
	ciphertext, err := hex.DecodeString(parts[1])
	if err != nil {
		return "", err
	}

	key := getReclaimEncKey(seed)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	plaintext, err := aesgcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return "", err
	}

	return string(plaintext), nil
}

func buildStatechainScript(sePkd string, reclaimTimelock int) string {
	kissHex := "0X" + strings.ToUpper(strings.TrimPrefix(sePkd, "0x"))
	return strings.Join([]string{
		"LET OWNER=STATE(0)",
		fmt.Sprintf("IF @COINAGE GTE %d THEN", reclaimTimelock),
		"  RETURN SIGNEDBY(OWNER)",
		"ENDIF",
		fmt.Sprintf("ASSERT MULTISIG(2 OWNER %s)", kissHex),
		"RETURN TRUE",
	}, "\n")
}

func scriptAddress(script string) string {
	h := sha3.New256()
	h.Write([]byte(strings.TrimSpace(strings.ToUpper(script))))
	return hex.EncodeToString(h.Sum(nil))
}
